Engineering/Concurrency in Javascript.md

```diff
@@ -0,0 +1,155 @@
+---
+tags: engineering, javascript, concurrency
+author: Hieu Phan
+---
+
+Javascript is a single-threaded programming language that has single-threaded at the runtime. It has a single call stack so it can do one thing simultaneously. JavaScript has a runtime model based on an **event loop**, which is responsible for executing the code, collecting and processing events, and executing queued sub-tasks. Javascript is a single-threaded, non-blocking asynchronous concurrent language. A JavaScript runtime uses a message queue, a list of messages to be processed. Each message has an associated function that gets called to handle the message.
+
+```javascript
+while (queue.waitForMessage()) {
+  queue.processNextMessage();
+}
+```
+Each message is processed completely before any other message is processed.
+There are three components in the runtime concepts: stack, heap, and queue.
+- Stack: contains the function calls. Each function call creates a stack frame that includes the statement and expression in the function. When that function returns a value or void, its frame is popped out, and the next function will begin executing. If we call a function inside another function, a frame will be created for each.
+- Heap: is a largely unstructured area of memory allocated to the objects in our code.
+- Callback queue: is a data structure, message queue, that stores a list of messages to be processed. Each message has an associated function that gets called to handle the message.
+
+```javascript
+function mult(a, b) {
+  return a*b;
+}
+function square(n) {
+  return mult(n, n);
+}
+function printSquare(n) {
+  let squared = square(n);
+  console.log(squared);
+}
+
+printSquare(4);
+```
+
+When running the above program, the Javascript will run the function step-by-step and add the function call to the `stack`. The first is: that the `printSquare()` function is added to the stack. The next functions are `square()`, `mult()`. The items in the `stack` will be executed orderly.
+
+## Javascript is a single-threaded language
+
+In reality, we always execute many tasks for a long time. These are opening a big file, downloading an image, or invoking a third-party service... We need to wait for the function call to be completed. How do we deal with the function call with the execution time too long? The Javascript runtime that is supported from the web browser, node.js. The web browser, or JS engines,... provides the APIs that help us run the task outside the `stack`. It means the long-time functions have already been finished in the stack, and the following function in the `stack` won't be blocked.
+
+```mermaid
```

Comment:
Not sure if it works on Obsidian

Reply:
![image](https://user-images.githubusercontent.com/1130103/186865708-6f5c26e0-3e4f-48ba-bd14-6c54647f82e4.png)
All good 👌